src/amp-react-selector.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import ReactCompatibleBaseElement from './react-compat-base-element.js';
+import {
+  AmpContext,
+} from './amp-context.js';
+import {Slot} from './slot.js';
+
+const {
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+  useState,
+} = preactHooks;
+
+
+export const AmpSelectorContext = preact.createContext();
+
+
+export function AmpSelector(props) {
+  const {multiple, value, defaultValue} = props;
+  const [selected, setSelectedState] = useState(
+    value ? [].concat(value) :
+    defaultValue ? [].concat(defaultValue) :
+    []);
+  const setSelectedRef = useRef(value => {
+    if (props.onChange) {
+      props.onChange({target: {value: multiple ? value : value[0]}});
+    }
+    setSelectedState(value);
+  });
+  return preact.createElement(
+    props.tagName || 'div',
+    {
+      ...props,
+    },
+    preact.createElement(
+      AmpSelectorContext.Provider,
+      {
+        ...props,
+        value: {
+          multiple,
+          // TBD: controlled values require override of properties.
+          selected: value ? [].concat(value) : selected,
+          setSelected: setSelectedRef.current,
+        },
+      }
+    )
+  );
+}
+
+AmpSelector.Option = function(props) {
+  const {option} = props;
+  const selectorContext = useContext(AmpSelectorContext);
+  const {selected, setSelected, multiple} = selectorContext;
+  const optionProps = {
+    ...props,
+    option,
+    selected: selected.includes(option),
+    onClick: () => selectOption(option, multiple, selected, setSelected),
```

Comment:
Why is it necessary to pass so may args back to the `Selector`? Shouldn't this all be possible because `selectOption` closes over the values?

Reply:
Premature optimization. I redid it with a single `selectOption` callback.